src/XMakeBuildEngine/Evaluation/ProjectParser.cs

```diff
@@ -368,6 +389,53 @@ private ProjectItemElement ParseProjectItemElement(XmlElementWithLocation elemen
             return item;
         }
 
+        private void CheckMetadataName(string name, out bool isKnownAttribute, out bool isValidMetadataNameInAttribute)
+        {
+            if (!XmlUtilities.IsValidElementName(name))
+            {
+                isKnownAttribute = false;
+                isValidMetadataNameInAttribute = false;
+                return;
+            }
+
+            for (int i = 0; i < s_knownAttributesOnItem.Length; i++)
+            {
+                //  Case insensitive comparison so that mis-capitalizing an attribute like Include or Exclude results in an easy to understand
+                //  error instead of unexpected behavior
+                if (name == s_knownAttributesOnItem[i])
+                {
+                    isKnownAttribute = true;
+                    isValidMetadataNameInAttribute = false;
+                    return;
+                }
+                else if (name.Equals(s_knownAttributesOnItem[i], StringComparison.OrdinalIgnoreCase))
+                {
+                    isKnownAttribute = false;
+                    isValidMetadataNameInAttribute = false;
+                    return;
+                }
+            }
+
+            //  Resorve attributes starting with underscores in case we need to add more built-in attributes later
```

Comment:
One concern: there's a common pattern of naming _properties_ with leading underscores as a clue to the user that they're internal implementation details. I don't think this is common for item metadata, though, so I don't think we'll be stepping on too many toes here.

The only other idea I have to make this even rarer is to take a page from the C++ book and use double underscores. Which is very ugly.

Thoughts, @AndyGerlicher @cdmihai @Sarabeth-Jaffe-Microsoft @jeffkl ?